File: Introduction_to_Robotics_Course/Chapter1/Lesson3/Chapter1_Lesson3.py

import numpy as np

def simulate_shared_autonomy(x0=2.0, k=1.5, alpha=0.7, T=5.0, dt=0.01):
    n_steps = int(T/dt)
    x = x0
    xs, us = [x], []

    for t in range(n_steps):
        u_A = -k * x                # autonomous stabilizing feedback
        u_H = 0.5*np.sin(2*np.pi*t*dt)  # example bounded human input
        u   = alpha*u_A + (1-alpha)*u_H

        x += dt*u                   # plant: x_dot = u
        xs.append(x)
        us.append(u)
    return np.array(xs), np.array(us)

xs, us = simulate_shared_autonomy()
print("Final state:", xs[-1])
      

File: Introduction_to_Robotics_Course/Chapter1/Lesson3/Chapter1_Lesson3.java

public class SharedAutonomy1D {
    public static void main(String[] args){
        double x = 2.0, k = 1.5, alpha = 0.7;
        double T = 5.0, dt = 0.01;
        int nSteps = (int)(T/dt);

        for(int i=0; i < nSteps; i++){
            double t = i*dt;
            double uA = -k*x;
            double uH = 0.5*Math.sin(2*Math.PI*t);
            double u  = alpha*uA + (1-alpha)*uH;

            x += dt*u; // x_dot = u
        }
        System.out.println("Final state: " + x);
    }
}
      